### ISSUE TITLE
How can I get the middle layer output value of the model?

### ISSUE THREAD

**Issue by XiaoJiQingZhou** [espnet/espnet#3204]
Sorry, I am a newbie in the field of speech recognition.
How can I get the output value of the softmax layer of the pre-trained models.

**Comment by XiaoJiQingZhou** [comment]
> We don't provide such specific functions.
> I think it would be better for you to implement it by yourself.

Thank you for your answer. What functions do I need to call to see the specific network structure of these pre training models?

**Comment by danoneata** [comment]
Maybe an alternative (more hands-on) approach of getting access to the output probabilities of an ASR model (that is, `p(tokens[t] | audio, tokens[:t])`) is something along these lines:
- Load the pretrained model ([as done in the `asr_inference.py` script](https://github.com/espnet/espnet/blob/master/espnet2/bin/asr_inference.py#L75-L77)); see the [ESPnet Model Zoo repository](https://github.com/espnet/espnet_model_zoo) for more information regarding pretrained models:
```python
from espnet2.tasks.asr import ASRTask
asr_model, asr_train_args = ASRTask.build_model_from_file(
    asr_train_config, asr_model_file, device
)
```
- Pass the speech through the encoder (see either [the `asr/espnet_model.py` script](https://github.com/espnet/espnet/blob/master/espnet2/asr/espnet_model.py#L135) or [the `asr_inference.py` script](https://github.com/espnet/espnet/blob/master/espnet2/bin/asr_inference.py#L197)):
```python
encoder_out, encoder_out_lens = asr_model.encode(speech, speech_lengths)
```
- Pass the encoder features through the decoder and supply also the auto-regressive context of tokens (see [the `asr/espnet_model.py` script](https://github.com/espnet/espnet/blob/master/espnet2/asr/espnet_model.py#L259-L261)):
```python
decoder_out, _ = asr_model.decoder(encoder_out, encoder_out_lens, ys_in_pad, ys_in_lens)
```
The `decoder_out` values correspond to pre-softmax features; so in order to obtain probabilities, you'll have to apply the softmax function.

As I'm also not very experienced with ESPnet's internals, Shinji please correct me if any of the above is wrong.

**Comment by sw005320** [comment]
Oh, yes. You're right.
I was actually checking the title of the issue "How can I get the middle layer output value of the model?" and answered it by guessing that the question wants to dump such features in the model.
I overlooked the question about "How can I get the output value of the softmax layer of the pre-trained models."

**Comment by XiaoJiQingZhou** [comment]
> Maybe an alternative (more hands-on) approach of getting access to the output probabilities of an ASR model (that is, `p(tokens[t] | audio, tokens[:t])`) is something along these lines:
> 
> * Load the pretrained model ([as done in the `asr_inference.py` script](https://github.com/espnet/espnet/blob/master/espnet2/bin/asr_inference.py#L75-L77)); see the [ESPnet Model Zoo repository](https://github.com/espnet/espnet_model_zoo) for more information regarding pretrained models:
> 
> ```python
> from espnet2.tasks.asr import ASRTask
> asr_model, asr_train_args = ASRTask.build_model_from_file(
>     asr_train_config, asr_model_file, device
> )
> ```
> 
> * Pass the speech through the encoder (see either [the `asr/espnet_model.py` script](https://github.com/espnet/espnet/blob/master/espnet2/asr/espnet_model.py#L135) or [the `asr_inference.py` script](https://github.com/espnet/espnet/blob/master/espnet2/bin/asr_inference.py#L197)):
> 
> ```python
> encoder_out, encoder_out_lens = asr_model.encode(speech, speech_lengths)
> ```
> 
> * Pass the encoder features through the decoder and supply also the auto-regressive context of tokens (see [the `asr/espnet_model.py` script](https://github.com/espnet/espnet/blob/master/espnet2/asr/espnet_model.py#L259-L261)):
> 
> ```python
> decoder_out, _ = asr_model.decoder(encoder_out, encoder_out_lens, ys_in_pad, ys_in_lens)
> ```
> 
> The `decoder_out` values correspond to pre-softmax features; so in order to obtain probabilities, you'll have to apply the softmax function.
> 
> As I'm also not very experienced with ESPnet's internals, Shinji please correct me if any of the above is wrong.

Thank you very much!
